### ISSUE TITLE
Encoding videos with a lossless compression

### ISSUE THREAD

**Issue by TBPorqui** [PolyMeilex/Neothesia#4]
Right now the recorder version uses a h264 profile which is incompatible with lossless compression https://trac.ffmpeg.org/wiki/Encode/H.264
This is quite bad as most people using that feature would want to edit these videos and thus will need to compress it again.

**Comment by PolyMeilex** [comment]
Thanks for the feedback. Quick disclaimer, my knowledge about video encoding is nearly none existent.

So in supplied link there is a `Lossless H.264` section that explains that setting `crf` to `0` will produce losses result, would this be sufficient enough?

Or maybe there are some codecs better suited for this use case? H.264 is used only because it was the only one implemented in [mpeg_encoder](https://github.com/PolyMeilex/mpeg_encoder_next) library before I forked it

**Comment by TBPorqui** [comment]
Yes it should work (at least according to this post https://video.stackexchange.com/a/18385 )

I don't know if the others are better my knowledge about video encoding is also poor.
Maybe sizewise they differ ? As a comparison, I used lagarith with Synthesia and it produced for 1:30 long videos ~800 MB files

**Comment by PolyMeilex** [comment]
Sure, so for now we'll go with h264. Tomorrow I'll do some research to find out how to access said `cfg` option from C API, and if that does not works out then I'll see what other codecs have to offer. My guess is that h264 will be fine, weight is not a big concern.

**Comment by PolyMeilex** [comment]
I managed to find a way to change crf and preset, and did some testing, and result are quite suspicious:
- Default: (`medium` + crf="-1.0")
  - Encoded 20208 frames (337s) in 356s (104MB)
- `medium` + crf="0.0"
  - Encoded 20208 frames (337s) in 398s (57MB)
- `veryslow` + crf="0.0"
  - Encoded 20208 frames (337s) in 655s (56.1MB)

**Comment by TBPorqui** [comment]
I actually tried converting the output video with x264 lossless and it gave me similar results.
I then converted that same video using huffyuv which is lossless (6.2 GB) and compared with the x264 lossless with the same commands from the stachexchange post and obtained the same results :
SSIM Y:1.000000 (inf) U:1.000000 (inf) V:1.000000 (inf) All:1.000000 (inf)

So even though it's weirdly small it seems to actually work correctly I guess?

**Comment by PolyMeilex** [comment]
Going from `104MB` to `57MB` by changing the encoding to lossless is quite weird, I mean yeah it increased the encode time, so it's definitely doing some additional work in the background, but it's still oddly small.
I assume it's working...

I exposed `crf` and `preset` settings to neothesia, so you can play around with those settings if you want:
[it's located here](https://github.com/PolyMeilex/Neothesia/blob/e67c1b30907095dfcaed241aab2d6dadafbfa4a5/src/main.rs#L174)
First arg is `crf`
Secong arg is `preset`
Setting them to `None` will simply use a default setting from `ffmpeg`

**Comment by TBPorqui** [comment]
Thanks!

As for now I'd say we leave it at that until an encoding expert stumble upon this project

**Comment by PolyMeilex** [comment]
For now it will stay as is. 
In the meantime I will experiment with other codecs, as with FFmpeg it should be quite straight forward to test out a few different ones. 
I'm trying to encode a YUV file as we speak (it is basically a sequence of images, it weights 62.9GB xddd), and I'll use this as a base for further testing

**Comment by PolyMeilex** [comment]
The proper way to fix this will be to add a proper CLI interface to the neothesia-cli, then encoding options can be just CLI flags.
